File: 백준/Gold/1753. 최단경로/최단경로.cc
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

/* 문제 분석
- 문제 설명
    - 방향 그래프 : 주어진 시작점에서 다른 모든 정점으로의 최단 경로 구하기
    - i번째 줄에 i번 정점으로의 최단 경로의 값 출력 ( 시작점은 자신은 0으로 출력, 경로가 없으면 INF 출력 )
- 입력 데이터
    - 정점의 개수 V
    - 간선의 개수 E
    - 시작 정점 K
    - 간선 + 가중치 : 모든 가중치는 10 이하의 가중치
- 핵심 키워드
    - 양의 가중치 & 최단 경로 -> 다익스트라
*/

void Dijkstra(const vector<vector<pair<int,int>>>& weight, vector<int>& distance, const int K){
    int n = weight.size();
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;  // {누적 가중치, 도착 정점}

    // 초기 세팅
    distance[K] = 0;
    pq.push({0, K});

    while(!pq.empty()){
        int wSum = pq.top().first;
        int start = pq.top().second;
        pq.pop();

        // 현재 경로가 기존에 찾은 경로보다 가중치가 작은 긴 경우 패스
        if(wSum > distance[start]){
            continue;
        }

        // 인근 경로 탐색
        for(const auto& [end, w] : weight[start]){
            int cost = distance[start] + w;
            if(cost < distance[end]){
                distance[end] = cost;
                pq.push({cost, end});
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    // 입력 데이터
    int V, E, K;
    cin >> V >> E >> K;

    vector<vector<pair<int,int>>> weight(V);  // 출발 정점 : {도착 정점, 가중치}
    for(int e = 0; e < E; e++){
        int start, end, w;
        cin >> start >> end >> w;
        weight[start-1].push_back({end-1, w});
    }

    // K부터 각 정점의 최단 경로 계산
    vector<int> distance(V, INT_MAX);  // 각 정점의 최단 경로    
    Dijkstra(weight, distance, K - 1);

    // 모든 최단 경로 출력
    for(int w : distance){
        if(w == INT_MAX){
            cout << "INF\n";
        } else{
            cout << w << "\n";
        }
    }
    
    return 0;
}

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Pair{
    int first;
    int second;

    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}

class Main {
    public static void Dijkstra(final List<List<Pair>> weight, ArrayList<Integer> distance, final int K){
        int n = weight.size();
        PriorityQueue<Pair> pq = new PriorityQueue<>(new Comparator<Pair>() {
            @Override
            public int compare(Pair p1, Pair p2) {
                // 누적 가중치(weight)가 작은 순으로 정렬
                return Integer.compare(p1.first, p2.first);
            }
        });
        // PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(p -> p.first));  // {누적 가중치, 도착 정점}

        // 초기 세팅
        distance.set(K, 0);
        pq.add(new Pair(0, K));

        while(!pq.isEmpty()){
            Pair current = pq.poll();
            int wSum = current.first;
            int start = current.second;

            // 현재 경로가 기존에 찾은 경로보다 가중치가 작은 경우 패스
            if(wSum > distance.get(start)){
                continue;
            }

            // 인근 경로 탐색
            for(Pair pair : weight.get(start)){
                int end = pair.first;
                int w = pair.second;
                int cost = distance.get(start) + w;
                if(cost < distance.get(end)){
                    distance.set(end, cost);
                    pq.offer(new Pair(cost, end));
                }
            }
        }
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // 입력 데이터
        int V = sc.nextInt();  // 정점의 개수
        int E = sc.nextInt();  // 간선의 개수
        int K = sc.nextInt() - 1;  // 시작 정점 (0-based index)


        // 가중치 리스트 초기화
        List<List<Pair>> weight = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            weight.add(new ArrayList<>());
        }

        // 간선 정보 입력
        for (int e = 0; e < E; e++) {
            int start = sc.nextInt() - 1;  // 시작 정점 (0-based index)
            int end = sc.nextInt() - 1;    // 끝 정점 (0-based index)
            int w = sc.nextInt();          // 가중치
            weight.get(start).add(new Pair(end, w));
        }

        // K부터 각 정점의 최단 경로 계산
        ArrayList<Integer> distance = new ArrayList<>(Collections.nCopies(V, Integer.MAX_VALUE));  // 각 정점의 최단 경로 (초기값은 INF)
        distance.set(K, 0);  // 시작 정점의 거리 초기화
        Dijkstra(weight, distance, K);

        // 모든 최단 경로 출력
        for (int w : distance) {
            if (w == Integer.MAX_VALUE) {
                System.out.println("INF");
            } else {
                System.out.println(w);
            }
        }
        
        sc.close();
    }
}